TECH_DEBT: Documentation Accuracy, Logging Improvements and Scripts

seanmcilvenna opened the pull request:
### 🛠️ Description of Changes
Adding error-checking and logging to ReportScheduledListener on Report service.
Adding producer for ReportScheduled to eventually use in UI. For now, beneficial to have operation in REST API to enable QA to bypass producing this event in Kafka UI.

### 🧪 Testing Performed
Please describe the testing that was performed on the changes included in this PR.

### 🧑‍🔬 Unit Testing
- [x] I have written or updated unit tests to cover my changes

### 📓 Documentation Updated

Updating docs for account and tenant to accurately reflect topics/events.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a script to analyze FHIR Bundle JSON files, providing resource statistics and identifying missing subject or patient references.
  - Added a script to detect missing CodeSystem URLs from validation response files, reporting unknown CodeSystems.

- **Bug Fixes**
  - Improved handling and logging in report scheduling to better manage report IDs and event processing.

- **Documentation**
  - Updated service documentation to clarify and expand listed Kafka events for Account and Tenant services.

- **Chores**
  - Enhanced test diagnostics with more detailed output during report submission status checks.
  - Updated Kafka broker configuration in the development environment for improved connectivity.

- **Refactor**
  - Improved logging and error handling for null events and duplicate report detection in report listeners.

- **Style**
  - Minor formatting adjustments in documentation for consistency.


<!-- end of auto-generated comment: release notes by coderabbit.ai -->

seanmcilvenna committed:
Updating docs for account and tenant to accurately reflect topics/events.
Adding error-checking and logging to ReportScheduledListener on Report service.

seanmcilvenna committed:
Adding producer for ReportScheduled to eventually use in UI. For now, beneficial to have operation in REST API to enable QA to bypass producing this event in Kafka UI.

nvmLantana commented on DotNet/Report/Listeners/ReportScheduledListener.cs:
```diff
@@ -105,18 +105,47 @@ await consumer.ConsumeWithInstrumentation(async (result, consumeCancellationToke
 
                                 ReportScheduleModel? reportSchedule;
-                                if(existing != null) 
+                                if (existing != null) 
```
We should just deadletter the message if the Report ID already exists. If you want to rerun a report, you can do so with the RegenerateReport option. This is not a feature that we wanted long term, and we have no legitimate use case for rerunning the exact same report Id again, and it has not been even been tested as a unique workflow.

seanmcilvenna replied:
I've changed it to produce a deadletter exception when existing != null

seanmcilvenna committed:
Fixing startup problem with Report service not having a producer registered correctly.
Instead of handling pre-existing report tracking ids in `ReportScheduledListener`, just producing a dead letter if it already exists. Already-existing reports are handled else-where.
Improving output of AdhocReportingSmokeTest

seanmcilvenna commented on DotNet/Report/KafkaProducers/ReportScheduledProducer.cs:
```diff
@@ -0,0 +1,73 @@
+            {
+                _logger.LogDebug(
+                    "Report scheduled event successfully produced. Report tracking ID: {ReportTrackingId}, Facility ID: {FacilityId}",
+                    reportTrackingId, facilityId);
```
I'm changing it to use the string extension `.SanitizeAndRemove()` for facilityId, and reportTrackingId.

seanmcilvenna commented on DotNet/Report/KafkaProducers/ReportScheduledProducer.cs:
```diff
@@ -0,0 +1,73 @@
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Error producing report scheduled event for facility {FacilityId}", facilityId);
```
I'm changing it to use the string extension .SanitizeAndRemove() for facilityId.

seanmcilvenna committed:
Sanitizing logs

seanmcilvenna committed:
Validation of input fields in `ScheduleReport()`

seanmcilvenna commented on DotNet/Report/KafkaProducers/ReportScheduledProducer.cs:
```diff
@@ -0,0 +1,74 @@
+    /// <param name="reportTypes">List of report types to generate</param>
+    /// <param name="frequency">The frequency of the report</param>
+    /// <returns>True if the message was successfully delivered to Kafka</returns>
+    public async Task<bool> Produce(
+        string reportTrackingId,
+        string facilityId,
+        DateTimeOffset startDate,
+        DateTimeOffset endDate,
+        List<string> reportTypes,
+        Frequency frequency)
```
Input validation is happening in the controller, now. Not going to duplicate that here.

nvmLantana reviewed (approved):
Looks Good!

nvmLantana committed:
Merge branch 'dev' into tech_debt/docs-reportscheduled-logging

seanmcilvenna committed:
Adding two useful scripts

seanmcilvenna committed:
Merge branch 'dev' into tech_debt/docs-reportscheduled-logging

seanmcilvenna committed:
Removing `ReportScheduledProducer` since equivalent functionality is already in the AdminBFF `IntegrationTesting` endpoint.

seanmcilvenna committed:
Merge remote-tracking branch 'origin/tech_debt/docs-reportscheduled-logging' into tech_debt/docs-reportscheduled-logging
